qiskit/pulse/ir/ir.py

```diff
@@ -0,0 +1,375 @@
+
+    @property
+    def initial_time(self) -> int | None:
+        """Return the initial time ``initial_time`` of the object"""
+        elements_initial_times = [element.initial_time for element in self._elements]
+        if None in elements_initial_times:
+            return None
+        else:
+            return min(elements_initial_times, default=None)
+
+    @property
+    def final_time(self) -> int | None:
+        """Return the final time of the ``IrBlock``object"""
+        elements_final_times = [element.final_time for element in self._elements]
+        if None in elements_final_times:
+            return None
+        else:
+            return max(elements_final_times, default=None)
+
+    @property
+    def duration(self) -> int | None:
+        """Return the duration of the ir block"""
+        try:
+            return self.final_time - self.initial_time
+        except TypeError:
+            return None
+
+    @property
+    def elements(self) -> List[IrElement]:
+        """Return the elements of the ``IrBlock`` object"""
+        return self._elements
+
+    @property
+    def alignment(self) -> AlignmentKind:
+        """Return the alignment of the ``IrBlock`` object"""
+        return self._alignment
+
+    def has_child_ir(self) -> bool:
+        """Check if IrBlock has child IrBlock object
+
+        Returns:
+            ``True`` if object has ``IrBlock`` object in its elements, and ``False`` otherwise.
+
+        """
+        for element in self._elements:
+            if isinstance(element, IrBlock):
+                return True
+        return False
+
+    def add_element(
+        self,
+        element: IrElement | List[IrElement],
+    ):
+        """Adds IR element or list thereof to the ``IrBlock`` object.
+
+        Args:
+            element: `IrElement` object, or list thereof to add.
+        """
+        if not isinstance(element, list):
+            element = [element]
+
+        self._elements.extend(element)
+
+    def shift_initial_time(self, value: int):
+        """Shifts ``initial_time`` of the ``IrBlock`` elements
+
+        The ``initial_time`` of all elements in the ``IrBlock`` are shifted by ``value``,
+        including recursively if any element is ``IrBlock``.
+
+        Args:
+            value: The integer value by which ``initial_time`` is to be shifted.
+
+        Raises:
+            PulseError: if the object is not scheduled (initial_time is None)
+        """
+        if self.initial_time is None:
+            raise PulseError("Can not shift initial_time of IrBlock with unscheduled elements")
+
+        for element in self.elements:
+            element.shift_initial_time(value)
+
+    def flatten(self) -> None:
+        """Flatten the ``IrBlock`` into a single block (in place)
+
+        Recursively flattens all child ``IrBlock`` until only instructions remain.
+
+        Raises:
+            PulseError: If the block is not scheduled.
+        """
+        if self.initial_time is None:
```

Comment:
You cannot flatten the nested IR with different alignment. I think this is more complicate operation and must be implemented by a pass.

Reply:
This is why I limited it to a scheduled IrBlock. At this point, you can safely flatten the representation. However, we can remove it for now, as it's not essential.